### DISCUSSION TITLE
Retrieving PathPattern from Endpoint

### DISCUSSION THREAD

**Discussion by rwnd-bradley** [poem-web/poem#663] [Q&A]
Hello, I am creating a rate limiter endpoint which is tracking based on request IP and URI. This works well except for paths with patterns (i.e. `/object/:id`). Instead of using `req.uri()` I'd like to use something similar to PathPattern, is this possible in middleware?

```rust
impl<E: Endpoint> Endpoint for RateLimitEndpoint<E> {
    type Output = Response;

    async fn call(&self, req: Request) -> Result<Self::Output> {
        let key = format!(
            "{}:{}",
            RealIp::from_request_without_body(&req)
                .await
                .map(|addr| addr.0.map(|addr| addr.to_string()))?
                .unwrap_or("localhost".to_string()),
            req.uri() // want this to be "/object/:id" instead of "/object/1", "/object/2", etc
        );

        // ..
    }
}
```

**Comment by miikka** [comment]
This issue comment indicates that the middleware cannot access the PathPattern data: https://github.com/poem-web/poem/issues/797#issuecomment-2039387365
